qualtran/bloqs/rotations/phase_gradient.py

```diff
@@ -302,15 +313,23 @@ def scaled_val(self, x: int) -> int:
     def apply(self, *args) -> Tuple[Union[int, np.integer, NDArray[np.integer]], ...]:
         if self.controlled_by is not None:
             ctrl, x, phase_grad = args
-            out = self.on_classical_vals(ctrl=ctrl, x=x, phase_grad=phase_grad)
-            return out['ctrl'], out['x'], out['phase_grad']
+            if ctrl == self.controlled_by:
+                phase_grad_out = (phase_grad + self.sign * self.scaled_val(x)) % (
+                    2**self.phase_bitsize
+                )
+            else:
+                phase_grad_out = phase_grad
+
+            return ctrl, x, phase_grad_out
 
         x, phase_grad = args
-        out = self.on_classical_vals(x=x, phase_grad=phase_grad)
-        return out['x'], out['phase_grad']
+        phase_grad_out = (phase_grad + self.sign * self.scaled_val(x)) % (2**self.phase_bitsize)
+        return x, phase_grad_out
+
+    def on_classical_vals(self, **kwargs) -> Dict[str, Union['ClassicalValT', Fxp]]:
+        x_fxp, phase_grad_fxp = kwargs['x'], kwargs['phase_grad']
 
-    def on_classical_vals(self, **kwargs) -> Dict[str, 'ClassicalValT']:
-        x, phase_grad = kwargs['x'], kwargs['phase_grad']
+        x, phase_grad = _extract_raw_int_from_fxp(x_fxp), _extract_raw_int_from_fxp(phase_grad_fxp)
```

Comment:
The `self.scaled_val(x)` logic was written because `on_classical_vals` expected an `int` instead of `Fxp`. Now that `on_classical_vals` expects an `Fxp` object directly, we can avoid converting `Fxp -> int -> scaled_int -> scaled_fxp` and directly do `Fxp -> Scaled Fxp` here. 

Maybe rename `self.scaled_val -> self._scaled_val_int` and call from `self.apply` and add a `self.scaled_val` that expects an Fxp object for `x` and returns the scaled value as 

```python
    def scaled_val(self, x: Fxp) -> Fxp:
        """Computes `phase_grad + x` using fixed point arithmetic."""
        return x.like(_fxp(0, self.phase_bitsize)) >> self.right_shift
```